File: _ui/src/lib/api/mcp-sets.ts
import axios from 'axios';
import type { ListResult, ListParams } from './types';
import type { MCPServerConfig, MCPStdioStatusResult } from './mcp-servers';

export type { MCPStdioStatusResult, MCPStdioUpstreamStatus } from './mcp-servers';

const api = axios.create({
  baseURL: 'api/v1',
});

// ─── Types ───

export interface MCPSet {
  id: string;
  owner_user_id?: string;
  scope?: 'personal' | 'workspace';
  name: string;
  description: string;
  category?: string;
  tags?: string[];
  config: MCPServerConfig;
  servers: string[];
  urls: string[];
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

// ─── CRUD ───

export async function listMCPSets(params?: ListParams): Promise<ListResult<MCPSet>> {
  const res = await api.get<ListResult<MCPSet>>('/mcp/sets', { params });
  return res.data;
}

export async function getMCPSet(id: string): Promise<MCPSet> {
  const res = await api.get<MCPSet>(`/mcp/sets/${id}`);
  return res.data;
}

export async function createMCPSet(data: Partial<MCPSet>): Promise<MCPSet> {
  const res = await api.post<MCPSet>('/mcp/sets', data);
  return res.data;
}

export async function updateMCPSet(id: string, data: Partial<MCPSet>): Promise<MCPSet> {
  const res = await api.put<MCPSet>(`/mcp/sets/${id}`, data);
  return res.data;
}

export async function deleteMCPSet(id: string): Promise<void> {
  await api.delete(`/mcp/sets/${id}`);
}

export async function publishMCPSet(id: string): Promise<MCPSet> {
  const res = await api.post<MCPSet>(`/mcp/sets/${id}/publish`);
  return res.data;
}

// ─── Import / Export ───

export async function exportMCPSet(id: string): Promise<Partial<MCPSet>> {
  const res = await api.get<Partial<MCPSet>>(`/mcp/sets/${id}/export`);
  return res.data;
}

export async function importMCPSet(data: Partial<MCPSet>): Promise<MCPSet> {
  const res = await api.post<MCPSet>('/mcp/sets/import', data);
  return res.data;
}

export async function previewImportMCPSet(data: Partial<MCPSet>): Promise<Partial<MCPSet>> {
  const res = await api.post<Partial<MCPSet>>('/mcp/sets/import/preview', data);
  return res.data;
}

// ─── Stdio upstream lifecycle ───

export async function getMCPSetStdioStatus(id: string): Promise<MCPStdioStatusResult> {
  const res = await api.get<MCPStdioStatusResult>(`/mcp/sets/${id}/stdio-status`);
  return res.data;
}

export async function restartMCPSetStdio(id: string, index?: number): Promise<MCPStdioStatusResult> {
  const body = index !== undefined ? { index } : {};
  const res = await api.post<MCPStdioStatusResult>(`/mcp/sets/${id}/stdio-restart`, body);
  return res.data;
}

export async function stopMCPSetStdio(id: string, index?: number): Promise<MCPStdioStatusResult> {
  const body = index !== undefined ? { index } : {};
  const res = await api.post<MCPStdioStatusResult>(`/mcp/sets/${id}/stdio-stop`, body);
  return res.data;
}

// ─── Tool Resolution (for Chat UI) ───

export interface MCPSetTool {
  name: string;
  description: string;
  inputSchema: Record<string, any>;
}

export interface MCPSetToolsResult {
  tools: MCPSetTool[];
  warnings?: string[];
}

export interface MCPSetToolCallResult {
  content: Array<{ type: string; text: string }>;
}

export interface MCPUpstreamInspection {
  index: number;
  transport: 'streamable_http' | 'stdio';
  response_mode?: 'json' | 'sse';
  protocol_version?: string;
  server_name?: string;
  server_version?: string;
  tool_count: number;
  tools: MCPSetTool[];
  duration_ms: number;
  error?: string;
}

export interface MCPUpstreamInspectionResult {
  name: string;
  upstreams: MCPUpstreamInspection[];
}

export async function inspectMCPSetUpstreams(id: string, index?: number): Promise<MCPUpstreamInspectionResult> {
  const res = await api.post<MCPUpstreamInspectionResult>(`/mcp/sets/${id}/inspect-upstreams`, index === undefined ? {} : { index });
  return res.data;
}

export async function listMCPSetTools(name: string): Promise<MCPSetToolsResult> {
  const res = await api.get<MCPSetToolsResult>(`/mcp/set-tools/${name}`);
  return res.data;
}

export async function callMCPSetTool(
  name: string,
  toolName: string,
  args: Record<string, any>,
): Promise<MCPSetToolCallResult> {
  const res = await api.post<MCPSetToolCallResult>(`/mcp/set-tools/${name}/call`, {
    tool_name: toolName,
    arguments: args,
  });
  return res.data;
}

File: internal/store/memory/mcp-sets.go
package memory

import (
	"context"
	"slices"
	"time"

	"github.com/oklog/ulid/v2"
	"github.com/rakunlabs/at/internal/service"
	"github.com/rakunlabs/query"
)

func (m *Memory) ListMCPSets(_ context.Context, q *query.Query) (*service.ListResult[service.MCPSet], error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	result := make([]service.MCPSet, 0, len(m.mcpSets))
	for _, s := range m.mcpSets {
		result = append(result, s)
	}

	slices.SortFunc(result, func(a, b service.MCPSet) int {
		if a.Name < b.Name {
			return -1
		}
		if a.Name > b.Name {
			return 1
		}
		return 0
	})

	return paginate(result, q), nil
}

func (m *Memory) GetMCPSet(_ context.Context, id string) (*service.MCPSet, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	s, ok := m.mcpSets[id]
	if !ok {
		return nil, nil
	}

	return &s, nil
}

func (m *Memory) GetMCPSetByName(_ context.Context, name string) (*service.MCPSet, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	for _, s := range m.mcpSets {
		if s.Name == name {
			return &s, nil
		}
	}

	return nil, nil
}

func (m *Memory) CreateMCPSet(_ context.Context, set service.MCPSet) (*service.MCPSet, error) {
	id := ulid.Make().String()
	now := time.Now().UTC().Format(time.RFC3339)

	rec := service.MCPSet{
		ID:          id,
		Name:        set.Name,
		Description: set.Description,
		Category:    set.Category,
		Tags:        set.Tags,
		Config:      set.Config,
		Servers:     set.Servers,
		URLs:        set.URLs,
		CreatedAt:   now,
		UpdatedAt:   now,
		CreatedBy:   set.CreatedBy,
		UpdatedBy:   set.UpdatedBy,
	}

	m.mu.Lock()
	m.mcpSets[id] = rec
	m.mu.Unlock()

	return &rec, nil
}

func (m *Memory) UpdateMCPSet(_ context.Context, id string, set service.MCPSet) (*service.MCPSet, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.mcpSets[id]
	if !ok {
		return nil, nil
	}

	existing.Name = set.Name
	existing.Description = set.Description
	existing.Category = set.Category
	existing.Tags = set.Tags
	existing.Config = set.Config
	existing.Servers = set.Servers
	existing.URLs = set.URLs
	existing.UpdatedAt = now
	existing.UpdatedBy = set.UpdatedBy
	m.mcpSets[id] = existing

	return &existing, nil
}

func (m *Memory) DeleteMCPSet(_ context.Context, id string) error {
	m.mu.Lock()
	delete(m.mcpSets, id)
	m.mu.Unlock()

	return nil
}
